### ISSUE TITLE
Source/destination can not be always populated

### ISSUE THREAD

**Issue by vbohata** [elastic/ecs#320]
Issue related to https://github.com/elastic/ecs/issues/300 and https://github.com/elastic/ecs/issues/246
I just found the log file / format where it is not possible to map client/server fields to source/destination. It is Exchange tracking log. It contains field "server-ip" which is "address of source or destination server". So while it makes sense to use server.ip field here, it does not make sense to copy it to destination.ip. This is the second time I found something like this in some log file. So I think copying client/server to source/destination should not be forced or even recommended in all cases.

**Comment by ave19** [comment]
I don't think ECS's goal is to make you fill out every field.

Advice:  If your data source only knows the "client/server" relationship, only fill out those fields.  If it only knows the "source/destination", then use those.  Leave the other ones blank.

There are some other issue threads like #9 and #63 which were related to that.

Hope this helps!

**Comment by webmat** [comment]
Thanks for the feedback. Yeah the relationship between src/dst and cli/srv will have to be fleshed out more completely. There's even some situations (e.g. auditd logs) where the kernel only tells you local/remote without a hint of which side initiated (the source). So I've found that these logs simply cannot be mapped to ECS at all.

Therefore perhaps we need the loc/rem pair as well. With a total of 3 pairs of endpoints, we definitely need to flesh out the "recommended way" and document such edge cases.

I'm even aware of aware of a situation where there's not even a possibility to map a network event to any of these 3 pairs: lossy network taps, where you may miss the SYN, but still observe part of the stream. In this situation:

- there is no concept of "local", it's a tap, independent from both endpoints
- there's no concept of "source", the SYN was missed
- there's no concept of "server", you just have two IP addresses and no context what their relationship is

At this point, do we create the `endpoint.a` and `endpoint.b` field set as well? 😂 Fun times ahead!

**Comment by vbohata** [comment]
Yes, and because of all these normalizations there is (and in future will be bigger) also issue with understanding of meaning the field names while comparing them to original log file field names. Especially in these (not always) edge cases. Also for "normal" cases where ECS mapping works there will have to be some improvements. Imagine following logfile->ECS mapping for myapp: notificationType->event.kind, resource_id->event.action (this is a real example).
So if the support of the myapp needs to know value of notificationType, the admin will be unable to find it...
So I think in future there will have to be:
* Support for alternate names in Elasticsearch so user could work with field event.kind the same way as with "my_system_original.notificationType" (not possible with the current alias implementation). The only way to do it now is to duplicate the data which is very space consuming.
* Support of renaming ECS field name at source (filebeat, winlogbeat, ...) so users will be able to make some configuration like "send event.id as my.eventid". This would make life much easier with keeping backward compatibility forever if required and avoiding collisions with myapp mapping.

**Comment by webmat** [comment]
Actually, to achieve the experience of having both an ECS field and the original naming for the same information, you can use the Elasticsearch `alias` field type.

I would highly recommend saving the data in the event _source with the ECS field name, and create as many aliases as you need, to allow the original naming of these fields to be used as well, without duplicating the data. This is the approach taken by Beats 7.0, to migrate to ECS, for example.

**Comment by vbohata** [comment]
Yes, but alias do not offer the same experience. For example the aliased fields are not visible in Kibana (there are not part of _source). Also not usable with objects.

**Comment by ave19** [comment]
For cases where we have field names coming from sources and we don't want to much around with them too much, we move the whole structure.  Like so:

1.  Create a field:  ` "event.type": "my_data"`
2.  Move your data into that object, so that you end up with:
```
{ 
   "event": {
      "type": "my_data"
   },
   "my_data": {
     "all": "the",
     "original": "fields",
     "source_ip": "1.2.3.4",
     "etc": "etc"
   },
   {
       "source": {
         "ip": "1.2.3.4"
     }  
}
```
Now, things like scripts can look for `event.type` and find the name of a top level structure, then pull all the original fields.  Then you can start copying `my_data` fields to their ECS destinations.

We do this so we don't LOSE the original message/structure.  Have and eat your cake.

hope this helps, I might not be completely tracking your goals. :smile:

**Comment by webmat** [comment]
I've made a note to come up with a solid documentation page explaining options, including pros & cons, on how to use aliases, or to make the duplication palatable.

I'm pretty sure we can dig out a few specific recommendations from the [Elasticsearch "How To" guides](https://www.elastic.co/guide/en/elasticsearch/reference/current/how-to.html) about performance, disk usage & so on.

As always, thanks for the feedback and ideas, guys 👍

**Comment by webmat** [comment]
@ave19 I like the approach you describe.

Although please be aware that `event.type` has been made reserved for the time being. We have something very specific planned for it soon, but not ready yet. I would recommend you use another field name for this technique.

**Comment by vbohata** [comment]
@ave19 We do something similar sometimes only if we need to rename some of our fields. In Logstash I copy some fields (object and numeric if changed units - seconds to ms ...), in other cases I use aliases.
Maybe the Kibana should solve it by itself - showing field alias names next to their field names in Discover, showing it in selections with italic font, ... This would help users with orientation.

**Comment by webmat** [comment]
Kibana 7 will have better support for aliases :-)

**Comment by ave19** [comment]
thanks for the heads up on event.type!

-ave
